### PR TITLE
chore: upgrading to design tokens v4

### LINKED ISSUE #24916: [EXT] Update design-tokens package to version 4.0.0
### **Description**

Upgrade the MetaMask extension to use design tokens from version 3 to version 4. This update involves numerous changes to color tokens. Follow the migration documentation provided in the [MetaMask Design Tokens Migration Guide](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#from-version-300-to-400).

### **Technical Details**

- Update all color tokens as per the migration documentation.
- Ensure compatibility with existing components and themes.
- Test thoroughly to ensure no visual regressions or inconsistencies are introduced.
- Follow the migration steps outlined in the [MetaMask Design Tokens Migration Guide](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#from-version-300-to-400).

### **Acceptance Criteria**

- All color tokens are updated to version 4.
- No visual regressions or inconsistencies in the extension UI.
- All relevant components and themes are verified to be compatible with the updated tokens.
- Documentation and comments are updated to reflect the changes.
- All tests pass, including any new tests added for the updated tokens.

### **References**

- [MetaMask Design Tokens Migration Guide](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#from-version-300-to-400)

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

This pull request aims to upgrade the extension to [design tokens v4](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#from-version-300-to-400). This upgrade ensures the most up-to-date colors are being used that align with design and primes the extension for the upcoming brand evolution. Included in this PR are third-party network color CSS variables that have been removed from the design tokens package and renamed CSS variables for shadow. This PR is the final update from a series of PRs that replace deprecated CSS colors that have been removed in v4.

**Dependency PRs that should be merged before this one:**
- https://github.com/MetaMask/metamask-extension/pull/24970
- https://github.com/MetaMask/metamask-extension/pull/24970
- https://github.com/MetaMask/metamask-extension/pull/25125
- https://github.com/MetaMask/metamask-extension/pull/25124
- https://github.com/MetaMask/metamask-extension/pull/25122
- https://github.com/MetaMask/metamask-extension/pull/25011
- https://github.com/MetaMask/metamask-extension/pull/25010
- https://github.com/MetaMask/metamask-extension/pull/25083
- https://github.com/MetaMask/metamask-extension/pull/24971
- https://github.com/MetaMask/metamask-extension/pull/25158

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24953?quickstart=1)

image

image (1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/24916

## **Manual testing steps**

Check all [removed and changed](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#from-version-300-to-400) CSS variables are no longer in the codebase:

1. Pull this branch.
2. Copy [this script](https://gist.github.com/georgewrmarshall/1ca0d7044bfa3c91343cadbf2bd5c828) that checks for all deprecated colors to the root of the extension.
3. Run `node searchDeprecatedTokens.js`.
4. Ensure there are no results.

Check shadows and network colors work as expected:

1. Navigate to stories that use primary and error shadows (`ButtonPrimary`, `ButtonSecondary`, `Button` (deprecated)) as well as network colors (`Box` BackgroundColors story).
2. Verify that shadows and network colors work as expected.
Run the extension and navigate around to ensure colors work as expected:

1. Pull this branch.
2. Run `yarn start`.
3. Navigate around the extension in light and dark mode to ensure colors work as expected.

### **Before**

Checking the codebase for any existing deprecated colors returns many results in the `develop` branch.

https://github.com/MetaMask/metamask-extension/assets/8112138/8ad23392-a0cc-4be5-8a50-12a126f5b71c

### **After**

Checking the codebase for any deprecated removed colors that could break the UI using the provided script, after removing network colors, returns no results.

https://github.com/MetaMask/metamask-extension/assets/8112138/3b82862d-5302-466f-bc55-bbb05a924ac0

Checking components that use updated CSS variables and newly added still work as expected

https://github.com/MetaMask/metamask-extension/assets/8112138/b70279b9-07f3-4315-abcd-83090ca03eef

Checking extension colors are working as expected in light and dark mode. In dark mode primary and error are a shade lighter in v4

https://github.com/MetaMask/metamask-extension/assets/8112138/d748f8cb-f2f2-4bd5-b9f4-f3bb376f22cd

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  const files = fs.readdirSync(directory);
+
+  files.forEach((file) => {
+    const filePath = path.join(directory, file);
+    const stat = fs.statSync(filePath);
+
+    if (stat.isDirectory()) {
+      log.push(...checkFiles(filePath));
+    } else if (filePath.endsWith('.css') || filePath.endsWith('.js')) {
+      const content = fs.readFileSync(filePath, 'utf-8');
```
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply script.js]: Fixed

File: .storybook/3.COLORS.stories.mdx [hunk 2]
```diff
@@ -74,49 +74,38 @@ var(--color-border-muted)
 
 /** Overlays */
 var(--color-overlay-default)
-var(--color-overlay-inverse) [DEPRECATED]
```
Comment (georgewrmarshall) [inline .storybook/3.COLORS.stories.mdx:77]: Not deprecated decided to keep as there are use cases for it
```suggestion
var(--color-overlay-inverse)
```

File: .storybook/3.COLORS.stories.mdx [hunk 3]
```diff
@@ -74,49 +74,38 @@ var(--color-border-muted)
...
 
 /** User Actions */
 var(--color-primary-default)
+var(--color-primary-default-hover)
+var(--color-primary-default-pressed)
```
Comment (georgewrmarshall) [inline .storybook/3.COLORS.stories.mdx:81]: Adding missing tokens to docs

File: .storybook/3.COLORS.stories.mdx (hunk 2 above)
Comment (georgewrmarshall) [reply .storybook/3.COLORS.stories.mdx:77]: fixed

File: .storybook/3.COLORS.stories.mdx [hunk 4]
```diff
@@ -123,7 +122,7 @@ Another level of abstraction is component tier colors that you can define at the
 
 ```scss
 .button {
-  --color-background-primary: var(--color-primary-default);
+  --color-background-success: var(--color-primary-default);
```
Comment (georgewrmarshall) [inline .storybook/3.COLORS.stories.mdx:125]: Revert this change
```suggestion
  --color-background-primary: var(--color-primary-default);
```

File: .storybook/4.SHADOW.stories.mdx [hunk 5]
```diff
@@ -115,7 +115,7 @@ As well as the neutral colors for shadow 2 other colors exist that are used for
     style={{
       height: 100,
       backgroundColor: 'var(--color-primary-default)',
-      boxShadow: 'var(--shadow-size-lg) var(--color-primary-shadow)',
+      boxShadow: 'var(--shadow-size-lg) var(--color-shadow-primary',
```
Comment (georgewrmarshall) [inline .storybook/4.SHADOW.stories.mdx:118]:
```suggestion
      boxShadow: 'var(--shadow-size-lg) var(--color-shadow-primary)',
```

File: .storybook/4.SHADOW.stories.mdx [hunk 6]
```diff
@@ -129,7 +129,7 @@ As well as the neutral colors for shadow 2 other colors exist that are used for
     style={{
       height: 100,
       backgroundColor: 'var(--color-error-default)',
-      boxShadow: 'var(--shadow-size-lg) var(--color-error-shadow)',
+      boxShadow: 'var(--shadow-size-lg) var(--color-shadow-error',
```
Comment (georgewrmarshall) [inline .storybook/4.SHADOW.stories.mdx:132]:
```suggestion
      boxShadow: 'var(--shadow-size-lg) var(--color-shadow-error)',
```

File: .storybook/4.SHADOW.stories.mdx [hunk 7]
```diff
@@ -144,8 +144,8 @@ As well as the neutral colors for shadow 2 other colors exist that are used for
 | Color       | CSS                           |
 | ----------- | ----------------------------- |
 | **neutral** | `var(--color-shadow-default)` |
-| **primary** | `var(--color-primary-shadow)` |
-| **danger**  | `var(--color-error-shadow)`   |
+| **primary** | `var(--color-shadow-primary)` |
+| **danger**  | `var(--color-shadow-error)`   |
```
Comment (georgewrmarshall) [inline .storybook/4.SHADOW.stories.mdx:148]: In [design tokens v4](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#css-1) `--color-primary-shadow` and `--color-error-shadow` where changed to `--color-shadow-primary` and `--color-shadow-error`. All instances have been replaced and primary and error shadows still work

Screenshot 2024-06-09 at 8 57 43 PM

Screenshot 2024-06-09 at 8 59 20 PM
Screenshot 2024-06-09 at 9 00 32 PM

Screenshot shows primary shadow still working after v4 upgrade

Screenshot 2024-06-09 at 9 06 17 PM
Screenshot 2024-06-09 at 9 07 06 PM

File: package.json [hunk 8]
```diff
@@ -290,7 +290,7 @@
     "@metamask/browser-passworder": "^4.3.0",
     "@metamask/contract-metadata": "^2.5.0",
     "@metamask/controller-utils": "^10.0.0",
-    "@metamask/design-tokens": "^3.0.0",
+    "@metamask/design-tokens": "^4.0.0",
     "@metamask/desktop": "^0.3.0",
```
Comment (georgewrmarshall) [inline package.json:294]: Upgrading version number in package.json

File: ui/css/utilities/colors.scss [hunk 9]
```diff
@@ -9,6 +9,10 @@ Before adding a color here make sure that there isn't a design token available.
   --mainnet: #29b6af;
   --inherit: inherit;
   --transparent: transparent;
+  --color-network-goerli-default: #1098fc;
+  --color-network-sepolia-default: #cfb5f0;
+  --color-network-goerli-inverse: #fcfcfc;
+  --color-network-sepolia-inverse: #fcfcfc;
```
Comment (georgewrmarshall) [inline ui/css/utilities/colors.scss:15]: These third party network colors have been removed from v4 of the design tokens. There was some previous thoughts that we could manage test network colors within the MetaMask brand as test networks generally don't have any branding elements attached, but ultimately they are third party entities and should be treated as such. We should treat test networks the same as other networks and use images (even if we are the ones creating those images). Eventually these should be replaced. 

From design tokens v4 [migration docs](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#removed)

Screenshot 2024-06-09 at 9 11 11 PM

 I have created a ticket to remove these colors here https://github.com/MetaMask/metamask-extension/issues/24969
 
 Move this to the design token v4 branch so we can ensure that they are still working in that PR
 
 Network colors still work after upgrade

Screenshot 2024-06-09 at 9 13 52 PM

File: ui/components/component-library/button-primary/button-primary.scss [hunk 10]
```diff
@@ -17,7 +17,7 @@
       color: var(--color-error-inverse);
       box-shadow:
         var(--shadow-size-sm)
-        var(--color-error-shadow);
+        var(--color-shadow-error);
     }
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-primary/button-primary.scss:21]: Screenshot shows primary shadow still working after v4 upgrade:

Screenshot 2024-06-09 at 9 06 17 PM
Screenshot 2024-06-09 at 9 07 06 PM

File: ui/components/component-library/button-secondary/button-secondary.scss [hunk 11]
```diff
@@ -24,7 +24,7 @@
       background-color: var(--color-error-default);
       box-shadow:
         var(--shadow-size-sm)
-        var(--color-error-shadow);
+        var(--color-shadow-error);
```
Comment (georgewrmarshall) [inline ui/components/component-library/button-secondary/button-secondary.scss:27]: Screenshot shows shadows still working after v4 upgrade:

Screenshot 2024-06-09 at 9 24 40 PM
Screenshot 2024-06-09 at 9 25 28 PM

File: ui/css/utilities/colors.scss [hunk 12]
```diff
@@ -9,13 +9,21 @@ Before adding a color here make sure that there isn't a design token available.
...
+  // Required for the QR reader to work properly
+  --qr-code-white-background: #fff;
+  // DEPRECATED
+  // These third party network colors have been deprecated and should be removed once they are no longer in use. We should be using images to represent these networks instead.
+  --color-network-goerli-default: #1098fc;
+  --color-network-sepolia-default: #cfb5f0;
+  --color-network-goerli-inverse: #fcfcfc;
+  --color-network-sepolia-inverse: #fcfcfc;
+  --color-network-localhost-default: #bbc0c5;
+  --color-network-localhost-inverse: #fcfcfc;
```
Comment (georgewrmarshall) [inline ui/css/utilities/colors.scss:22]: These third party network colors have been removed from v4 of the design tokens. There was some previous thoughts that we could manage test network colors within the MetaMask brand as test networks generally don't have any branding elements attached, but ultimately they are third party entities and should be treated as such. We should treat test networks the same as other networks and use images (even if we are the ones creating those images). Eventually these should be replaced. 

From design tokens v4 [migration docs](https://github.com/MetaMask/design-tokens/blob/main/MIGRATION.md#removed)

Screenshot 2024-06-09 at 9 11 11 PM

I have created a ticket to remove these colors, as they are third-party colors, not widely used, and unlikely to affect the brand evolution. It's low priority: https://github.com/MetaMask/metamask-extension/issues/24969
 
 Move this to the design token v4 branch so we can ensure that they are still working in that PR
 
 Network colors still work after upgrade

Screenshot 2024-06-09 at 9 13 52 PM

File: ui/css/utilities/colors.scss [hunk 13]
```diff
@@ -9,13 +9,21 @@ Before adding a color here make sure that there isn't a design token available.
...
   --color-network-linea-goerli-default: #61dfff;
   --color-network-linea-goerli-inverse: #fcfcfc;
   --color-network-linea-sepolia-default: #61dfff;
   --color-network-linea-sepolia-inverse: #fcfcfc;
   --color-network-linea-mainnet-default: #121212;
   --color-network-linea-mainnet-inverse: #fcfcfc;
-  // DO NOT CHANGE
-  // Required for the QR reader to work properly
-  --qr-code-white-background: #fff;
```
Comment (georgewrmarshall) [inline ui/css/utilities/colors.scss:20]: Moved above deprecated network colors

File: ui/css/utilities/colors.scss [hunk 14]
```diff
@@ -9,13 +9,21 @@ Before adding a color here make sure that there isn't a design token available.
   --mainnet: #29b6af;
   --inherit: inherit;
   --transparent: transparent;
+  // DO NOT CHANGE
+  // Required for the QR reader to work properly
+  --qr-code-white-background: #fff;
```
Comment (georgewrmarshall) [inline ui/css/utilities/colors.scss:14]: Moved up from bottom of the lists

File: ui/components/ui/button/buttons.scss [hunk 15]
```diff
@@ -262,7 +262,7 @@ input[type="submit"][disabled] {
     background-color: var(--color-error-default);
 
     &:hover {
-      box-shadow: var(--shadow-size-sm) var(--color-error-shadow);
+      box-shadow: var(--shadow-size-sm) var(--color-shadow-error);
     }
```
Comment (georgewrmarshall) [inline ui/components/ui/button/buttons.scss:266]: ### After

Shadows still work as expected in deprecated button component


https://github.com/MetaMask/metamask-extension/assets/8112138/2e328890-cd2f-40f7-a46b-f8cc5dd692e8

File: .storybook/3.COLORS.stories.mdx [hunk 16]
```diff
@@ -18,7 +18,7 @@ We follow a 3 tiered system for color design tokens and css variables.
 <div
   style={{
     textAlign: 'center',
-    backgroundColor: 'var(--brand-colors-white-white000)',
+    backgroundColor: 'var(--brand-colors-white)',
```
Comment (georgewrmarshall) [inline .storybook/3.COLORS.stories.mdx:21]: ### After

Background color to image still works as expected


https://github.com/MetaMask/metamask-extension/assets/8112138/5d112637-07db-40ac-95ae-575691e40209

File: .storybook/3.COLORS.stories.mdx [hunk 17]
```diff
@@ -36,9 +36,9 @@ These colors **SHOULD NOT** be used in your styles directly. They are used as a
 
 ```css
 /** !!!DO NOT USE BRAND COLORS DIRECTLY IN YOUR CODE!!! */
-var(--brand-colors-white-white000)
-var(--brand-colors-white-white010)
-var(--brand-colors-grey-grey030)
+var(--brand-colors-white)
+var(--brand-colors-black)
+var(--brand-colors-grey-grey800)
```
Comment (georgewrmarshall) [inline .storybook/3.COLORS.stories.mdx:41]: Updating docs to reflect correct tokens

File: .storybook/3.COLORS.stories.mdx [hunk 18]
```diff
@@ -168,7 +168,7 @@ Don't use static hex values or brand color tokens in your code.
 * Not theme compatible and will break UI when using dark theme
 **/
 .card {
-  background-color: var(--brand-colors-white-white000);
+  background-color: var(--brand-colors-white);
```
Comment (georgewrmarshall) [inline .storybook/3.COLORS.stories.mdx:171]: Updating to correct token in docs

File: ui/components/multichain/address-copy-button/index.scss [hunk 19]
```diff
@@ -8,7 +8,7 @@
 
 .custody-logo {
   height: 20px;
-  background-color: var(--brand-colors-white-white000);
+  background-color: var(--brand-colors-white);
   padding: 1px;
```
Comment (georgewrmarshall) [inline ui/components/multichain/address-copy-button/index.scss:12]: ### After
Background of custody logo is white

Screenshot 2024-06-10 at 3 13 42 PM
Screenshot 2024-06-10 at 3 13 46 PM

File: ui/components/multichain/ramps-card/index.scss [hunk 20]
```diff
@@ -4,15 +4,15 @@
...
 
   &__title {
-    color: var(--brand-colors-white-white000);
+    color: var(--brand-colors-white);
   }
 
   &__body {
-    color: var(--brand-colors-white-white000);
+    color: var(--brand-colors-white);
     width: 80%;
```
Comment (georgewrmarshall) [inline ui/components/multichain/ramps-card/index.scss:16]: ### After
Ramps card text and button remain white

Screenshot 2024-06-10 at 3 15 58 PM
Screenshot 2024-06-10 at 3 16 19 PM

File: ui/css/base-styles.scss [hunk 21]
```diff
@@ -31,7 +31,7 @@ html {
   */
   @media (prefers-color-scheme: dark) {
     &:not([data-theme]) {
-      color: var(--brand-colors-white-white000);
+      color: var(--brand-colors-white);
```
Comment (georgewrmarshall) [inline ui/css/base-styles.scss:34]: This seems like the default font color for prefers color scheme dark. Check before/after screenshot of description as settings for that recording are set to system

File: ui/pages/unlock-page/index.scss [hunk 22]
```diff
@@ -25,7 +25,7 @@
     &__beta {
       /* these colors should be used on both light and dark mode */
       background: var(--brand-colors-blue-blue500);
-      color: var(--brand-colors-white-white000);
+      color: var(--brand-colors-white);
       padding: 3px 6px;
```
Comment (georgewrmarshall) [inline ui/pages/unlock-page/index.scss:29]: ### After
Beta text remains white

Screenshot 2024-06-10 at 3 26 13 PM

Comment (brianacnguyen) [reply ui/pages/unlock-page/index.scss:29]: should this be using a themed token instead?

**Review by brianacnguyen — COMMENTED** [review]
1 small comment

File: ui/pages/unlock-page/index.scss (hunk 22 above)
Comment (georgewrmarshall) [reply ui/pages/unlock-page/index.scss:29]: No, see comment on line 26 `/* these colors should be used on both light and dark mode */`

Comment (georgewrmarshall) [reply ui/pages/unlock-page/index.scss:28]: Good question! I think originally it used brand colors here because the designer wanted the text to remain white but it would look strange once we update the colors so I'll update to theme. Updated here https://github.com/MetaMask/metamask-extension/pull/24953/commits/ddaeba1ec3fbbbf6e4895f09c4ac0864f7d14d9f

Screenshot 2024-06-10 at 4 28 09 PM
Screenshot 2024-06-10 at 4 28 35 PM

Comment (brianacnguyen) [reply ui/pages/unlock-page/index.scss:28]: nice!

**Review by brianacnguyen — DISMISSED** [review]

**Review by garrettbear — DISMISSED** [review]

**Comment by bergeron** [comment]
Thanks for annotating the changes so it's easier to review

**Review by garrettbear — APPROVED** [review]

**Review by brianacnguyen — APPROVED** [review]
